apiserver/tools.go

```diff
@@ -277,98 +278,31 @@ func (h *toolsDownloadHandler) getToolsForRequest(r *http.Request) (_ io.ReadClo
 	if err != nil {
 		return nil, 0, errors.Trace(err)
 	}
-	binaries, err := ds.AgentBinary().ListAgentBinaries(r.Context())
-	if err != nil {
-		return nil, 0, errors.Trace(err)
-	}
-	sha256 := ""
-	for _, v := range binaries {
-		// TODO(agentbinary): perform this search without listing all cached
-		// agent binaries.
-		if v.Version == vers.Number.String() && v.Arch == vers.Arch {
-			sha256 = v.SHA256
-			break
-		}
-	}
-	if sha256 == "" {
-		// TODO(agentbinary): if the binary is not found, ask simplestreams for
-		// it, then cache it in the model agent binary store. It is possible to
-		// cache in the controller agent binary store if the source stream is
-		// the default.
-		return nil, 0, errors.NotFound
-	}
 
-	// Check the current model's agent binary store first.
-	f, size, err := ds.AgentBinaryStore().GetAgentBinaryForSHA256(r.Context(), sha256)
-	if errors.Is(err, agentbinaryerrors.NotFound) {
-		// Fallback to the controller agent binary store.
-		f, size, err = ds.ControllerAgentBinaryStore().GetAgentBinaryForSHA256(r.Context(), sha256)
+	ctx := r.Context()
+	agentBinaryService := ds.AgentBinary()
+	var agentBinaryStream io.ReadCloser
+	var size int64
+
+	ver := coreagentbinary.Version{
+		Arch:   vers.Arch,
+		Number: vers.Number,
 	}
+
+	hasAgentBinary, err := agentBinaryService.HasAgentBinary(ctx, ver)
 	if err != nil {
-		return nil, 0, errors.Trace(err)
+		return nil, 0, errors.Errorf("checking if agent binary exists: %v", err)
 	}
 
-	return f, size, nil
-}
+	str := coreagentbinary.AgentStream(stream)
```

Comment:
Ah sorry @CodingCookieRookie, think we may have just got our wires cross here. We don't need the stream at all.

When a client hits this endpoint it is one of our agents running out in the model that is looking to download the latest tools available. Mostly likely to provision a new machine or perform an upgrade.

The stream in this context will always be that which the model is set to be running. Agents never get to decide this for them selves.

In affect we can remove this value and change `GetExternalAgentBinary` to look up what the stream should be from the current model.